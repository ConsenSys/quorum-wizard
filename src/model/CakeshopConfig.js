import {
  isDocker,
  isTessera,
} from './NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function generateCakeshopConfig(config) {
  const nodes = []

  config.nodes.forEach((node, i) => {
    const nodeData = {
      name: `node${i + 1}`,
      rpcUrl: isDocker(config.network.deployment)
        // docker-compose creates dns entries for the container's ip using container name (nodeX)
        ? `http://node${i + 1}:${config.containerPorts.quorum.rpcPort}`
        : `http://localhost:${node.quorum.rpcPort}`,
      transactionManagerUrl: createTransactionManagerUrl(config, node, i),
    }
    nodes.push(nodeData)
  })
  return nodes
}

function createTransactionManagerUrl(config, node, i) {
  if (isTessera(config.network.transactionManager)) {
    return isDocker(config.network.deployment)
      // docker-compose creates dns entries for the container's ip using container name (txmanagerX)
      ? `http://txmanager${i + 1}:${config.containerPorts.tm.thirdPartyPort}`
      : `http://localhost:${node.tm.thirdPartyPort}`
  }
  return ''
}
