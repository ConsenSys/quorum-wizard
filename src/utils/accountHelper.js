import { Wallet } from 'ethers'

export default function nodekeyToAccount(nodekey) {
  return new Wallet(nodekey).address
}
