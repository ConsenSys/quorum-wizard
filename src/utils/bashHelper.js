import { createDirectory, isTessera, networkFolder, includeCakeshop } from './networkCreator'
import {
  copyFile,
  createFolder,
  readFileToString,
  removeFolder,
  writeFile,
  writeJsonFile,
} from './fileUtils'
import { join, normalize } from 'path'
import { execute } from './execUtils'
import { buildCakeshopDir, generateCakeshopScript, waitForCakeshopCommand } from './cakeshopHelper'

export function buildBashScript(config) {
  const commands = createDirectory(config)

  const cakeshopStart = includeCakeshop(config) ? [generateCakeshopScript(), waitForCakeshopCommand()].join("") : ""

  const startScript = [
    commands.tesseraStart,
    waitForTesseraNodesCommand(config),
    commands.gethStart,
    cakeshopStart
  ]

  return {
    startScript: startScript.join('\n'),
    initCommands: commands.initStart,
    networkPath: commands.netPath
  }
}

export function buildBash(config) {

  const bashDetails = buildBashScript(config)
  const networkPath = bashDetails.networkPath

  if(includeCakeshop(config)) {
    buildCakeshopDir(config, join(networkPath, 'qdata'))
  }

  writeFile(join(networkPath, 'start.sh'), bashDetails.startScript, true)
  copyFile(join(process.cwd(), 'lib/stop.sh'), join(networkPath, 'stop.sh'))

  copyFile(join(process.cwd(), 'lib/runscript.sh'), join(networkPath, 'runscript.sh'))
  copyFile(join(process.cwd(), 'lib/public-contract.js'), join(networkPath, 'public-contract.js'))
  copyFile(join(process.cwd(), 'lib/private-contract.js'), join(networkPath, 'private-contract.js'))

  bashDetails.initCommands.forEach((command) => {
    // TODO figure out the safest way to run shell commands
    execute(command, (e, stdout, stderr) => {
      if (e instanceof Error) {
        console.error(e)
        throw e
      }
    })
  })
}

export function createGethStartCommand (config, node, passwordDestination, nodeNumber, tmIpcLocation) {
  const { verbosity, id, consensus } = config.network
  const { devP2pPort, rpcPort, wsPort, raftPort } = node.quorum

  const args = `--nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,${consensus},quorumPermission --emitcheckpoints --unlock 0 --password ${passwordDestination}`
  const consensusArgs = consensus === 'raft' ?
    `--raft --raftport ${raftPort}` :
    `--istanbul.blockperiod 5 --syncmode full --mine --minerthreads 1`

  return `PRIVATE_CONFIG=${tmIpcLocation} nohup geth --datadir qdata/dd${nodeNumber} ${args} ${consensusArgs} --permissioned --verbosity ${verbosity} --networkid ${id} --rpcport ${rpcPort} --port ${devP2pPort} 2>>qdata/logs/${nodeNumber}.log &`
}

export function createTesseraStartCommand (config, node, nodeNumber, tmDir, logDir) {
  // `rm -f ${tmDir}/tm.ipc`

  const tesseraJar = '$TESSERA_JAR' // require env variable to be set for now
  let DEBUG = ''
  if (config.network.remoteDebug) {
    DEBUG = '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=500$i -Xdebug'
  }

  const MEMORY = '-Xms128M -Xmx128M'
  const CMD = `java ${DEBUG} ${MEMORY} -jar ${tesseraJar} -configfile ${tmDir}/tessera-config-09-${nodeNumber}.json >> ${logDir}/tessera${nodeNumber}.log 2>&1 &`
  return CMD
}

export function waitForTesseraNodesCommand (config) {
  if(!isTessera(config)) {
    return ''
  }
  // TODO use config values for ip, port, data folder, etc.
  return `
echo "Waiting until all Tessera nodes are running..."
DOWN=true
k=10
while \${DOWN}; do
    sleep 1
    DOWN=false
    for i in \`seq 1 ${config.nodes.length}\`
    do
        if [ ! -S "qdata/c\${i}/tm.ipc" ]; then
            echo "Node \${i} is not yet listening on tm.ipc"
            DOWN=true
        fi

        set +e
        #NOTE: if using https, change the scheme
        #NOTE: if using the IP whitelist, change the host to an allowed host
        result=$(curl -s http://localhost:900\${i}/upcheck)
        set -e
        if [ ! "\${result}" == "I'm up!" ]; then
            echo "Node \${i} is not yet listening on http"
            DOWN=true
        fi
    done

    k=$((k - 1))
    if [ \${k} -le 0 ]; then
        echo "Tessera is taking a long time to start.  Look at the Tessera logs in qdata/logs/ for help diagnosing the problem."
    fi
    echo "Waiting until all Tessera nodes are running..."

    sleep 5
done

echo "All Tessera nodes started"
`
}
