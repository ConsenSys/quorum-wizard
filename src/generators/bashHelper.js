import {
  getFullNetworkPath,
} from './networkCreator'
import {
  copyFile,
  libRootDir,
  writeFile,
} from '../utils/fileUtils'
import { executeSync } from '../utils/execUtils'
import {
  buildCakeshopDir,
  generateCakeshopScript,
} from './cakeshopHelper'
import {
  isQuorum260Plus,
  pathToCakeshop,
  pathToQuorumBinary,
  pathToTesseraJar,
} from './binaryHelper'
import {
  isRaft,
  isTessera,
  isCakeshop,
} from '../model/NetworkConfig'
import { info } from '../utils/log'
import { formatTesseraKeysOutput } from './transactionManager'
import { joinPath } from '../utils/pathUtils'

export function buildBashScript(config) {
  const commands = createCommands(config)

  const startScript = [
    '#!/bin/bash',
    'echo "\nStarting Quorum network...\n"',
    setEnvironmentCommand(config),
    commands.tesseraStart,
    waitForTesseraNodesCommand(config),
    'echo "Starting Quorum nodes"',
    commands.gethStart,
    generateCakeshopScript(config),
    'echo "Successfully started Quorum network."',
    `echo "${formatTesseraKeysOutput(config)}"`,
  ]

  return {
    startScript: startScript.join('\n'),
    initCommands: commands.initStart,
  }
}

export function createCommands(config) {
  const networkPath = getFullNetworkPath(config)
  const initCommands = []
  const startCommands = []
  const tmStartCommands = []

  config.nodes.forEach((node, i) => {
    const nodeNumber = i + 1
    const quorumDir = joinPath('qdata', `dd${nodeNumber}`)
    const tmDir = joinPath('qdata', `c${nodeNumber}`)
    const genesisLocation = joinPath(quorumDir, 'genesis.json')
    const keyDir = joinPath(quorumDir, 'keystore')
    const passwordDestination = joinPath(keyDir, 'password.txt')
    const logs = joinPath('qdata', 'logs')
    const initCommand = `cd ${networkPath} && ${pathToQuorumBinary(config.network.quorumVersion)} --datadir ${quorumDir} init ${genesisLocation} 2>&1`
    initCommands.push(initCommand)

    const tmIpcLocation = isTessera(config.network.transactionManager)
      ? joinPath(tmDir, 'tm.ipc')
      : 'ignore'
    const startCommand = createGethStartCommand(
      config,
      node,
      passwordDestination,
      nodeNumber,
      tmIpcLocation,
    )
    startCommands.push(startCommand)

    if (isTessera(config.network.transactionManager)) {
      const tmStartCommand = createTesseraStartCommand(config, node, nodeNumber, tmDir, logs)
      tmStartCommands.push(tmStartCommand)
    }
  })

  const obj = {
    tesseraStart: tmStartCommands.join('\n'),
    gethStart: startCommands.join('\n'),
    initStart: initCommands,
  }
  return obj
}


export async function buildBash(config) {
  const bashDetails = buildBashScript(config)
  const networkPath = getFullNetworkPath(config)

  if (isCakeshop(config.network.cakeshop)) {
    buildCakeshopDir(config, joinPath(networkPath, 'qdata'))
  }

  info('Writing start script...')
  writeFile(joinPath(networkPath, 'start.sh'), bashDetails.startScript, true)
  copyFile(joinPath(libRootDir(), 'lib', 'stop.sh'), joinPath(networkPath, 'stop.sh'))

  info('Initializing quorum...')
  bashDetails.initCommands.forEach((command) => executeSync(command))
  info('Done')
}

export function createGethStartCommand(config, node, passwordDestination, nodeNumber, tmIpcPath) {
  const {
    verbosity, networkId, consensus, quorumVersion,
  } = config.network
  const {
    devP2pPort, rpcPort, raftPort, wsPort,
  } = node.quorum
  const quorum26Flags = isQuorum260Plus(quorumVersion) ? '--allow-insecure-unlock' : ''

  const args = `--nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,${consensus},quorumPermission --ws --wsaddr 0.0.0.0 --wsorigins=* --emitcheckpoints --unlock 0 --password ${passwordDestination} ${quorum26Flags}`
  const consensusArgs = isRaft(consensus)
    ? `--raft --raftport ${raftPort}`
    : '--istanbul.blockperiod 5 --syncmode full --mine --minerthreads 1'

  return `PRIVATE_CONFIG=${tmIpcPath} nohup $BIN_GETH --datadir qdata/dd${nodeNumber} ${args} ${consensusArgs} --permissioned --verbosity ${verbosity} --networkid ${networkId} --rpcport ${rpcPort} --wsport ${wsPort} --port ${devP2pPort} 2>>qdata/logs/${nodeNumber}.log &`
}

export function createTesseraStartCommand(config, node, nodeNumber, tmDir, logDir) {
  // `rm -f ${tmDir}/tm.ipc`

  let DEBUG = ''
  if (config.network.remoteDebug) {
    DEBUG = '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=500$i -Xdebug'
  }

  const MEMORY = '-Xms128M -Xmx128M'
  const CMD = `java ${DEBUG} ${MEMORY} -jar $BIN_TESSERA -configfile ${tmDir}/tessera-config-09-${nodeNumber}.json >> ${logDir}/tessera${nodeNumber}.log 2>&1 &`
  return CMD
}

function checkTesseraUpcheck(nodes) {
  return nodes.map((node, i) => `
    result${i + 1}=$(curl -s http://${node.tm.ip}:${node.tm.p2pPort}/upcheck)
    if [ ! "\${result${i + 1}}" == "I'm up!" ]; then
        echo "Node ${i + 1} is not yet listening on http"
        DOWN=true
    fi`)
}

export function setEnvironmentCommand(config) {
  const lines = []
  lines.push(`BIN_GETH=${pathToQuorumBinary(config.network.quorumVersion)}`)
  if (isTessera(config.network.transactionManager)) {
    lines.push(`BIN_TESSERA=${pathToTesseraJar(config.network.transactionManager)}`)
  }
  if (isCakeshop(config.network.cakeshop)) {
    lines.push(`BIN_CAKESHOP=${pathToCakeshop(config.network.cakeshop)}`)
  }
  lines.push('')
  return lines.join('\n')
}

export function waitForTesseraNodesCommand(config) {
  if (!isTessera(config.network.transactionManager)) {
    return ''
  }
  return `
echo "Waiting until all Tessera nodes are running..."
DOWN=true
k=10
while \${DOWN}; do
    sleep 1
    DOWN=false
    for i in \`seq 1 ${config.nodes.length}\`
    do
        if [ ! -S "qdata/c\${i}/tm.ipc" ]; then
            echo "Node \${i} is not yet listening on tm.ipc"
            DOWN=true
        fi
    done
    set +e
    #NOTE: if using https, change the scheme
    #NOTE: if using the IP whitelist, change the host to an allowed host
    ${checkTesseraUpcheck(config.nodes).join('')}

    k=$((k - 1))
    if [ \${k} -le 0 ]; then
        echo "Tessera is taking a long time to start.  Look at the Tessera logs in qdata/logs/ for help diagnosing the problem."
    fi
    echo "Waiting until all Tessera nodes are running..."

    sleep 5
done

echo "All Tessera nodes started"
`
}
