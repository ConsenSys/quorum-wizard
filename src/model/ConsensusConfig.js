import {
  writeFile,
  writeJsonFile,
  combineFiles,
  readFileToString
} from '../utils/fileUtils'
import { join } from 'path'
import { executeSync } from '../utils/execUtils'
const Web3 = require('web3');

export function generateConsensusConfig(configDir, consensus, nodes) {
  return consensus == `raft` ?  generateRaft(configDir, nodes) :
  generateIstanbul(configDir, nodes)
}

function generateRaft(configDir, nodes) {
  writeJsonFile(configDir, `raft-genesis.json`, generateRaftConfig(nodes, configDir))
}

function generateIstanbul(configDir, nodes) {
  writeJsonFile(configDir, `istanbul-genesis.json`, generateIstanbulConfig(nodes, configDir))
}

export function generateAccounts(nodes, configDir) {
  var numNodes = nodes.length
  var raftJsonString = `{`
  for (let i = 1; i < parseInt(numNodes, 10); i++) {
    let keyDir = join(configDir, `key${i}`)
    let keyString = readFileToString(join(keyDir, 'key'))

    raftJsonString += `\"${JSON.parse(keyString).address}\":{\"balance\":\"1000000000000000000000000000\"},`
  }
  let keyDir = join(configDir, `key${numNodes}`)
  let keyString = readFileToString(join(keyDir, 'key'))

  raftJsonString += `\"${JSON.parse(keyString).address}\":{\"balance\":\"1000000000000000000000000000\"}}`
  return JSON.parse(raftJsonString)
}

function generateRaftConfig(numNodes, configDir) {
  const alloc = generateAccounts(numNodes, configDir)
  return {
    alloc: alloc,
    coinbase: `0x0000000000000000000000000000000000000000`,
    config: {
      homesteadBlock: 0,
      byzantiumBlock: 0,
      constantinopleBlock: 0,
      chainId: 10,
      eip150Block: 0,
      eip155Block: 0,
      eip150Hash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
      eip158Block: 0,
      maxCodeSize: 35,
      isQuorum: true,
    },
    difficulty: `0x0`,
    extraData: `0x0000000000000000000000000000000000000000000000000000000000000000`,
    gasLimit: `0xE0000000`,
    mixhash: `0x00000000000000000000000000000000000000647572616c65787365646c6578`,
    nonce: `0x0`,
    parentHash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
    timestamp: `0x00`
  }
}

function generateExtraData(nodes, configDir) {
  var validatorConfigString = `vanity = \"0x00\"\nvalidators=[`
  var numNodes = nodes.length
  for (let i = 1; i < parseInt(numNodes, 10); i++) {
      let keyDir = join(configDir, `key${i}`)
      let nodekey = nodekeyToAccount(readFileToString(join(keyDir, 'nodekey')))
      validatorConfigString += `\"${nodekey}\",`
  }
  let keyDir = join(configDir, `key${numNodes}`)
  let nodekey = nodekeyToAccount(readFileToString(join(keyDir, 'nodekey')))
  validatorConfigString += `\"${nodekey}\"]`

  var istanbulConfigFile = join(configDir, 'istanbul.toml')
  writeFile(istanbulConfigFile, validatorConfigString, false)

  var extraDataCmd = `cd ${configDir} && istanbul extra encode --config ${istanbulConfigFile} | awk '{print $4}' > extraData`
  executeSync(extraDataCmd, function(err, stdout, stderr) {
    if (e instanceof Error) {
      console.error(e)
      throw e
    }
  })
  return readFileToString(join(configDir, 'extraData'))
}

function nodekeyToAccount(nodekey) {
  const web3 = new Web3();
  var acctObj = web3.eth.accounts.privateKeyToAccount(nodekey);
  return acctObj.address
}

function generateIstanbulConfig(nodes, configDir) {
  const alloc = generateAccounts(nodes, configDir)
  const extraData = generateExtraData(nodes, configDir)
  return {
    alloc: alloc,
    coinbase: `0x0000000000000000000000000000000000000000`,
    config: {
      homesteadBlock: 0,
      byzantiumBlock: 0,
      constantinopleBlock: 0,
      chainId: 10,
      eip150Block: 0,
      eip155Block: 0,
      eip150Hash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
      eip158Block: 0,
      maxCodeSize: 35,
      isQuorum: true,
      istanbul: {
        epoch: 30000,
        policy: 0,
        ceil2Nby3Block: 0,
      }
    },
    difficulty: `0x1`,
    extraData: extraData,
    gasLimit: `0xE0000000`,
    mixhash: `0x63746963616c2062797a616e74696e65206661756c7420746f6c6572616e6365`,
    nonce: `0x0`,
    parentHash: `0x0000000000000000000000000000000000000000000000000000000000000000`,
    timestamp: `0x00`
  }

}
