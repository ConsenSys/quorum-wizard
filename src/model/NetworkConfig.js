import { isJava11Plus } from '../utils/execUtils'
import {
  LATEST_CAKESHOP,
  LATEST_CAKESHOP_J8,
  LATEST_QUORUM,
  LATEST_TESSERA,
  LATEST_TESSERA_J8,
} from '../generators/download'
import {
  getDockerSubnet,
  cidrhost,
} from '../utils/subnetUtils'
import { getOutputPath } from '../utils/fileUtils'

export function createConfigFromAnswers(answers) {
  const {
    name,
    networkPath = getOutputPath(),
    numberNodes = 3,
    consensus = 'raft',
    quorumVersion = LATEST_QUORUM,
    transactionManager = isJava11Plus() ? LATEST_TESSERA : LATEST_TESSERA_J8,
    deployment = 'bash',
    cakeshop = isJava11Plus() ? LATEST_CAKESHOP : LATEST_CAKESHOP_J8,
    generateKeys = false,
    networkId = '10',
    genesisLocation = 'none',
    customizePorts = false,
    nodes = [],
    cakeshopPort = '8999',
    remoteDebug = false,
    containerPorts = undefined,
  } = answers
  const networkFolder = name
    || defaultNetworkName(numberNodes, consensus, transactionManager, deployment)
  const dockerSubnet = (isDocker(deployment) && containerPorts !== undefined) ? containerPorts.dockerSubnet : ''
  return {
    network: {
      name: networkFolder,
      networkPath,
      verbosity: 5,
      consensus,
      quorumVersion,
      transactionManager,
      permissioned: true,
      genesisFile: genesisLocation,
      generateKeys,
      configDir: `network/${networkFolder}/resources`,
      deployment,
      cakeshop,
      networkId,
      customizePorts,
      cakeshopPort,
      remoteDebug,
    },
    nodes: (customizePorts && nodes.length > 0) ? nodes : generateNodeConfigs(
      numberNodes,
      transactionManager,
      deployment,
      cakeshop,
      dockerSubnet,
    ),
    containerPorts,
  }
}

export function defaultNetworkName(numberNodes, consensus, transactionManager, deployment) {
  const transactionManagerName = !isTessera(transactionManager)
    ? ''
    : 'tessera-'
  return `${numberNodes}-nodes-${consensus}-${transactionManagerName}${deployment}`
}

export function generateNodeConfigs(
  numberNodes,
  transactionManager,
  deployment,
  cakeshop,
  dockerSubnet,
) {
  const devP2pPort = 21000
  const rpcPort = 22000
  const wsPort = 23000
  const graphQlPort = 24000
  const raftPort = 50401
  const thirdPartyPort = 9081
  const p2pPort = 9001
  const nodes = []

  for (let i = 0; i < parseInt(numberNodes, 10); i += 1) {
    const node = {
      quorum: {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 10) : '127.0.0.1',
        devP2pPort: devP2pPort + i,
        rpcPort: rpcPort + i,
        wsPort: wsPort + i,
        raftPort: raftPort + i,
        graphQlPort: graphQlPort + i,
      },
    }
    if (isTessera(transactionManager)) {
      node.tm = {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 100) : '127.0.0.1',
        thirdPartyPort: thirdPartyPort + i,
        p2pPort: p2pPort + i,
      }
    }
    nodes.push(node)
  }
  return nodes
}

export function getContainerPorts(deployment) {
  const dockerSubnet = isDocker(deployment) ? getDockerSubnet() : ''
  return {
    dockerSubnet,
    quorum: {
      rpcPort: 8545,
      p2pPort: 21000,
      raftPort: 50400,
      wsPort: 8645,
      graphQlPort: 8547,
    },
    tm: {
      p2pPort: 9000,
      thirdPartyPort: 9080,
    },
  }
}

export function isTessera(tessera) {
  return tessera !== 'none'
}

export function isDocker(deployment) {
  return deployment === 'docker-compose'
}

export function isBash(deployment) {
  return deployment === 'bash'
}

export function isKubernetes(deployment) {
  return deployment === 'kubernetes'
}

export function isIstanbul(consensus) {
  return consensus === 'istanbul'
}

export function isRaft(consensus) {
  return consensus === 'raft'
}

export function isCakeshop(cakeshop) {
  return cakeshop !== 'none'
}
