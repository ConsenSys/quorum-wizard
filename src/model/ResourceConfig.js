import { isTessera } from './NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function buildKubernetesResource(config) {
  return [
    buildGeneralDetails(config),
    buildQuorumDetails(config),
    buildTesseraDetails(config),
    buildGethDetails(config),
  ].join('')
}

function buildGeneralDetails(config) {
  const nodeNum = config.nodes.length
  return `#namespace:
#  name: quorum-test
sep_deployment_files: true
generate_keys: ${config.network.generateKeys}
nodes:
  number: ${nodeNum}
service:
  type: NodePort`
}

function buildQuorumDetails(config) {
  return `
quorum:
  consensus: ${config.network.consensus}
  Node_DataDir: /etc/quorum/qdata
  Key_Dir_Base: out/config
  Permissioned_Nodes_File: out/config/permissioned-nodes.json
  Genesis_File: out/config/genesis.json
  quorum:
    Raft_Port: ${config.nodes[0].quorum.raftPort}
    Quorum_Version: ${config.network.quorumVersion}
  storage:
    Type: PVC
    Capacity: 200Mi`
}

function buildGethDetails(config) {
  return `
geth:
  Node_RPCPort: 8545
  NodeP2P_ListenAddr: 30303
  network:
    id: ${config.network.networkId}
    public: false
  verbosity: 9
  Geth_Startup_Params: --rpccorsdomain=\\"*\\"`
}

function buildTesseraDetails(config) {
  const hasTessera = isTessera(config.network.transactionManager)
  return hasTessera
    ? `
  tm:
    Name: tessera
    Tm_Version: ${config.network.transactionManager}
    Port: ${config.nodes[0].tm.thirdPartyPort}
    Tessera_Config_Dir: out/config`
    : ''
}
