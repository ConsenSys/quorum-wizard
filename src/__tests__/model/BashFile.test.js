import { buildBashScript } from '../../utils/bashHelper'
import { createQuickstartConfig } from '../../model/NetworkConfig'

test('creates 3nodes raft bash tessera', () => {
  const config = createQuickstartConfig('3', 'raft', 'tessera', 'bash')
  const bash = buildBashScript(config).startScript
  expect(bash).toMatchSnapshot()
})
