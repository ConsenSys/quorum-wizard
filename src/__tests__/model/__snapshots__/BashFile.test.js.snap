// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates 3nodes raft bash tessera 1`] = `
"java  -Xms128M -Xmx128M -jar $TESSERA_JAR -configfile /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/c1/tessera-config-09-1.json >> /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/logs/tessera1.log 2>&1 &
java  -Xms128M -Xmx128M -jar $TESSERA_JAR -configfile /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/c2/tessera-config-09-2.json >> /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/logs/tessera2.log 2>&1 &
java  -Xms128M -Xmx128M -jar $TESSERA_JAR -configfile /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/c3/tessera-config-09-3.json >> /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/logs/tessera3.log 2>&1 &

echo \\"Waiting until all Tessera nodes are running...\\"
DOWN=true
k=10
while \${DOWN}; do
    sleep 1
    DOWN=false
    for i in \`seq 1 3\`
    do
        if [ ! -S \\"qdata/c\${i}/tm.ipc\\" ]; then
            echo \\"Node \${i} is not yet listening on tm.ipc\\"
            DOWN=true
        fi

        set +e
        #NOTE: if using https, change the scheme
        #NOTE: if using the IP whitelist, change the host to an allowed host
        result=$(curl -s http://localhost:900\${i}/upcheck)
        set -e
        if [ ! \\"\${result}\\" == \\"I'm up!\\" ]; then
            echo \\"Node \${i} is not yet listening on http\\"
            DOWN=true
        fi
    done

    k=$((k - 1))
    if [ \${k} -le 0 ]; then
        echo \\"Tessera is taking a long time to start.  Look at the Tessera logs in qdata/logs/ for help diagnosing the problem.\\"
    fi
    echo \\"Waiting until all Tessera nodes are running...\\"

    sleep 5
done

echo \\"All Tessera nodes started\\"

PRIVATE_CONFIG=/Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/c1/tm.ipc nohup geth --datadir qdata/dd1 --nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,raft,quorumPermission --emitcheckpoints --unlock 0 --password /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/dd1/keystore/password.txt --raft --raftport 50401 --permissioned --verbosity 5 --networkid 10 --rpcport 22000 --port 21000 2>>qdata/logs/1.log &
PRIVATE_CONFIG=/Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/c2/tm.ipc nohup geth --datadir qdata/dd2 --nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,raft,quorumPermission --emitcheckpoints --unlock 0 --password /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/dd2/keystore/password.txt --raft --raftport 50402 --permissioned --verbosity 5 --networkid 10 --rpcport 22001 --port 21001 2>>qdata/logs/2.log &
PRIVATE_CONFIG=/Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/c3/tm.ipc nohup geth --datadir qdata/dd3 --nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,raft,quorumPermission --emitcheckpoints --unlock 0 --password /Users/angelapratt/projects/quorum-creator/network/3-nodes-raft-tessera-bash/qdata/dd3/keystore/password.txt --raft --raftport 50403 --permissioned --verbosity 5 --networkid 10 --rpcport 22002 --port 21002 2>>qdata/logs/3.log &"
`;
